include/slam_toolbox/laser_utils.hpp

```diff
@@ -75,8 +76,9 @@ class LaserMetadata
 class LaserAssistant
 {
 public:
+  template<class NodeT>
```

Comment:
This is already templated by NodeT, why not have the NodeT stored instead of pulling out the logger / parameters_interface?

Reply:
> have the NodeT stored

The current template function must be converted to template class.   

The update will look like this.

<details> <summary> (minimum example) convert template function to template class of LaserAssistant </summary>

```diff
$ git diff
diff --git a/include/slam_toolbox/laser_utils.hpp b/include/slam_toolbox/laser_utils.hpp
index f7f81e6..ef5091d 100644
--- a/include/slam_toolbox/laser_utils.hpp
+++ b/include/slam_toolbox/laser_utils.hpp
@@ -73,12 +73,13 @@ private:
 };
 
 // Help take a scan from a laser and create a laser object
+template<class NodeT = rclcpp::Node>
 class LaserAssistant
 {
 public:
-  template<class NodeT>
+  typedef std::shared_ptr<NodeT> NodePtr;
   LaserAssistant(
-    NodeT node, tf2_ros::Buffer * tf,
+    NodePtr node, tf2_ros::Buffer * tf,
     const std::string & base_frame);
   ~LaserAssistant();
   LaserMetadata toLaserMetadata(sensor_msgs::msg::LaserScan scan);
@@ -87,6 +88,7 @@ private:
   karto::LaserRangeFinder * makeLaser(const double & mountingYaw);
   bool isInverted(double & mountingYaw);
 
+  NodePtr node_;
   rclcpp::Logger logger_;
   rclcpp::node_interfaces::NodeParametersInterface::SharedPtr parameters_interface_;
   tf2_ros::Buffer * tf_;
diff --git a/include/slam_toolbox/slam_toolbox_common.hpp b/include/slam_toolbox/slam_toolbox_common.hpp
index 85eb236..3e907f2 100644
--- a/include/slam_toolbox/slam_toolbox_common.hpp
+++ b/include/slam_toolbox/slam_toolbox_common.hpp
@@ -167,7 +167,7 @@ protected:
   std::map<std::string, laser_utils::LaserMetadata> lasers_;
 
   // helpers
-  std::unique_ptr<laser_utils::LaserAssistant> laser_assistant_;
+  std::unique_ptr<laser_utils::LaserAssistant<rclcpp_lifecycle::LifecycleNode>> laser_assistant_;
   std::unique_ptr<pose_utils::GetPoseHelper> pose_helper_;
   std::unique_ptr<map_saver::MapSaver> map_saver_;
   std::unique_ptr<loop_closure_assistant::LoopClosureAssistant> closure_assistant_;
diff --git a/src/laser_utils.cpp b/src/laser_utils.cpp
index fb54ef0..0cc3d4d 100644
--- a/src/laser_utils.cpp
+++ b/src/laser_utils.cpp
@@ -69,19 +69,21 @@ void LaserMetadata::invertScan(sensor_msgs::msg::LaserScan & scan) const
 }
 
 template<class NodeT>
-LaserAssistant::LaserAssistant(
-  NodeT node,
+LaserAssistant<NodeT>::LaserAssistant(
+  NodePtr node,
   tf2_ros::Buffer * tf, const std::string & base_frame)
 : logger_(node->get_logger()), parameters_interface_(node->get_node_parameters_interface()),
-  tf_(tf), base_frame_(base_frame)
+  tf_(tf), base_frame_(base_frame), node_(node)
 {
 }
 
-LaserAssistant::~LaserAssistant()
+template<class NodeT>
+LaserAssistant<NodeT>::~LaserAssistant()
 {
 }
 
-LaserMetadata LaserAssistant::toLaserMetadata(sensor_msgs::msg::LaserScan scan)
+template<class NodeT>
+LaserMetadata LaserAssistant<NodeT>::toLaserMetadata(sensor_msgs::msg::LaserScan scan)
 {
   scan_ = scan;
   frame_ = scan_.header.frame_id;
@@ -93,7 +95,8 @@ LaserMetadata LaserAssistant::toLaserMetadata(sensor_msgs::msg::LaserScan scan)
   return laserMeta;
 }
 
-karto::LaserRangeFinder * LaserAssistant::makeLaser(const double & mountingYaw)
+template<class NodeT>
+karto::LaserRangeFinder * LaserAssistant<NodeT>::makeLaser(const double & mountingYaw)
 {
   karto::LaserRangeFinder * laser =
     karto::LaserRangeFinder::CreateLaserRangeFinder(
@@ -147,7 +150,8 @@ karto::LaserRangeFinder * LaserAssistant::makeLaser(const double & mountingYaw)
   return laser;
 }
 
-bool LaserAssistant::isInverted(double & mountingYaw)
+template<class NodeT>
+bool LaserAssistant<NodeT>::isInverted(double & mountingYaw)
 {
   geometry_msgs::msg::TransformStamped laser_ident;
   laser_ident.header.stamp = scan_.header.stamp;
@@ -206,9 +210,6 @@ void ScanHolder::addScan(const sensor_msgs::msg::LaserScan scan)
 }
 
 // explicit instantiation for the supported template types
-template LaserAssistant::LaserAssistant(
-  rclcpp::Node::SharedPtr, tf2_ros::Buffer *, const std::string &);
-template LaserAssistant::LaserAssistant(
-  rclcpp_lifecycle::LifecycleNode::SharedPtr, tf2_ros::Buffer *, const std::string &);
-
+template class LaserAssistant<rclcpp::Node>;
+template class LaserAssistant<rclcpp_lifecycle::LifecycleNode>;
 }  // namespace laser_utils
diff --git a/src/slam_toolbox_common.cpp b/src/slam_toolbox_common.cpp
index 44b4ae1..46fbe1c 100644
--- a/src/slam_toolbox_common.cpp
+++ b/src/slam_toolbox_common.cpp
@@ -60,7 +60,7 @@ CallbackReturn SlamToolbox::on_configure(const rclcpp_lifecycle::State &)
   setROSInterfaces();
   setSolver();
 
-  laser_assistant_ = std::make_unique<laser_utils::LaserAssistant>(
+  laser_assistant_ = std::make_unique<laser_utils::LaserAssistant<rclcpp_lifecycle::LifecycleNode>>(
     shared_from_this(), tf_.get(), base_frame_);
   pose_helper_ = std::make_unique<pose_utils::GetPoseHelper>(
     tf_.get(), base_frame_, odom_frame_);
```

</details>

Reply:
Just leave it then